### PR TITLE
Fix dataset factory patterns in Experiment Tracking

### LINKED ISSUE #1480: No experiment tracking for datasets defined with a dataset factory
## Description
Experiment tracking does not detect datasets when defined with a dataset factory.

## Context
I've been using the recent [dataset factory](https://docs.kedro.org/en/stable/data/data_catalog.html#load-multiple-datasets-with-similar-configuration-using-dataset-factories) feature and it's been wonderful!

However, when I tried to visualize the tracked dataset with Kedro Viz experiment tracking feature, no tracked dataset is detected.

## Steps to Reproduce

1. Create a node that outputs some metrics dict structure under the `my_namespace.metrics` name
2. Update your catalog with a dataset factory:

```yaml
"{namespace}.metrics":
  type: tracking.MetricsDataSet
  filepath: data/{namespace}.metrics.json
```

3. Run your kedro pipeline, the file should now be saved under `data/my_namespace.metrics.json/<SESSION_ID>/my_namespace.metrics.json`
4. Run `kedro viz` and go to the latest experiment, under the "Overview" tab

## Expected Result
The expected behavior would be for the `my_namespace.metrics` JSON content to appear on the UI.

## Actual Result
There is no JSON content on the UI.

## Your Environment
Include as many relevant details as possible about the environment you experienced the bug in:

* Web browser system and version: Google Chrome - Version 115.0.5790.170 (Official Build) (64-bit)
* Operating system and version: Ubuntu 22.04.3 LTS (64bits)
* NodeJS version used (if relevant): -
* Kedro version used: 0.18.12
* Kedro Viz version used: 6.3.4
* Python version used: 3.9.17

## Checklist
- [x] Include labels so that we can categorise your issue

### PR DESCRIPTION
**Pull request by ravi-kumar-pilla**
## Description

Resolves #1480 

## Development notes

* Discovering datasets defined via dataset factory pattern using datasets in pipelines before adding catalog and tracking datasets to kedro viz
* Inspired from the PR - #1491 

Example patterns -

Screenshot 2023-10-18 at 1 11 46 PM

## QA notes

* [Steps to Reproduce](https://github.com/kedro-org/kedro-viz/issues/1480)
* You can also follow the dataset factory patterns [guide](https://docs.kedro.org/en/stable/data/kedro_dataset_factories.html#kedro-dataset-factories) and modify the demo-project or create a spaceflights project
* After executing `kedro run`, you should see the factory pattern datasets in the experiment tracking run

## Checklist

- [x] Read the [contributing](/CONTRIBUTING.md) guidelines
- [ ] Opened this PR as a 'Draft Pull Request' if it is work-in-progress
- [ ] Updated the documentation to reflect the code changes
- [x] Added new entries to the `RELEASE.md` file
- [x] Added tests to cover my changes

### CODE REVIEW

File: package/kedro_viz/data_access/managers.py [hunk 1]
```diff
@@ -82,6 +82,22 @@ def add_catalog(self, catalog: DataCatalog):
...
 
+    def resolve_dataset_factory_patterns(
+        self, catalog: DataCatalog, pipelines: Dict[str, KedroPipeline]
+    ):
+        """Resolve dataset factory patterns in data catalog by matching
+        them against the datasets in the pipelines.
+        """
+        for pipeline in pipelines.values():
+            # Temporary try/except block so the Kedro develop branch can work with Viz.
+            try:
```
Comment (ankatiyar) [inline package/kedro_viz/data_access/managers.py:93]: This might catch other exceptions. You could try `if hasattr(pipeline, 'data_set`):`

**Comment by noklam** [comment]
Is this ready to be reviewed or still in draft?

**Comment by ravi-kumar-pilla** [comment]
> Is this ready to be reviewed or still in draft?

I reverted it to draft PR in favor of [1491](https://github.com/kedro-org/kedro-viz/pull/1491) . Apologies for the confusion

**Comment by ravi-kumar-pilla** [comment]
> Is this ready to be reviewed or still in draft?

Hi @noklam, this is now ready for review

File: package/kedro_viz/data_access/managers.py [hunk 2]
```diff
@@ -69,6 +69,21 @@ def set_db_session(self, db_session_class: sessionmaker):
         """Set db session on repositories that need it."""
         self.runs.set_db_session(db_session_class)
 
+    def resolve_dataset_factory_patterns(
+        self, catalog: DataCatalog, pipelines: Dict[str, KedroPipeline]
```
Comment (noklam) [inline package/kedro_viz/data_access/managers.py:73]:
```suggestion
        self, catalog: DataCatalog, pipelines: Dict[str, Pipeline]
```

Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/managers.py:73]: Hi Nok, we have an import with alias as - `from kedro.pipeline import Pipeline as KedroPipeline` . This is used in other places as well, so followed the same.

File: package/kedro_viz/data_access/managers.py [hunk 3]
```diff
@@ -69,6 +69,21 @@ def set_db_session(self, db_session_class: sessionmaker):
...
+    ):
+        """Resolve dataset factory patterns in data catalog by matching
+        them against the datasets in the pipelines.
+        """
+        for pipeline in pipelines.values():
+            if hasattr(pipeline, "datasets"):
+                datasets = pipeline.datasets()
```
[the remark this answers is not in the document]
Comment (ravi-kumar-pilla) [reply package/kedro_viz/data_access/managers.py:80]: yup, sure will do

**Review by noklam — APPROVED** [review]

**Comment by noklam** [comment]
For reference, @rashidakanchwala shared an alternative solution [here](https://github.com/kedro-org/kedro-viz/compare/main...fix/dataset-factories-with-exp-tracking)

**Comment by ravi-kumar-pilla** [comment]
Hi @noklam , Thanks for mentioning it here. @rashidakanchwala and I had a discussion on this. So, based on what I understood from [this](https://github.com/kedro-org/kedro-viz/pull/964) PR -

1. Separation of responsibility - TrackingDatasetRepository was introduced and is populated during the add_catalog phase which is primarily used in experiment tracking runs. (can also be used in pipelines)
2. Experiment Tracking should not depend on creation of the DAG, rather it is a historical data of kedro runs. Ideally, ET data should not depend on even the catalog information as it is a tracker of historic data and the catalog can change between runs. But, as of now ET does not store data/path to the datasets. This makes it depend on the catalog information. 
4. This [PR](https://github.com/kedro-org/kedro-viz/compare/main...fix/dataset-factories-with-exp-tracking) has the tracking datasets being created during the `add_dataset` function (i.e., during the creation of node repos for DAG). This makes ET data to be dependant more on the creation of pipelines, which will not serve the purpose of separation of responsibility. 

Let me know what do you think of this. 

Thank you

**Pull request merged**